zebrad/src/components/mempool/storage/eviction_list.rs

```diff
@@ -0,0 +1,140 @@
+        // > and/or just before RecentlyEvicted is accessed when receiving a transaction.
+        self.prune_old();
+        // > Add the txid and the current time to RecentlyEvicted, dropping the oldest entry
+        // > in RecentlyEvicted if necessary to keep it to at most eviction_memory_entries entries.
+        if self.len() >= self.max_size {
+            self.pop_front();
+        }
+        let value = Instant::now();
+        let old_value = self.unique_entries.insert(key, value);
+        // It should be impossible for an already-evicted transaction to be evicted
+        // again since transactions are not added to the mempool if they are evicted,
+        // and the mempool doesn't allow inserting two transactions with the same
+        // hash (they would conflict).
+        assert!(
+            old_value.is_none(),
+            "an already-evicted transaction should not be evicted again"
+        );
+        self.ordered_entries.push_back(key)
+    }
+
+    /// Checks if the given TXID is in the list.
+    pub fn contains_key(&self, txid: &transaction::Hash) -> bool {
+        if let Some(evicted_at) = self.unique_entries.get(txid) {
+            // Since the list is pruned only in mutable functions, make sure
+            // we take expired items into account.
+            if !self.has_expired(evicted_at) {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Get the size of the list.
+    //
+    // Note: if this method being mutable becomes an issue, it's possible
+    // to compute the number of expired transactions and subtract.
+    pub fn len(&mut self) -> usize {
+        self.prune_old();
+        self.unique_entries.len()
+    }
+
+    /// Clear the list.
+    #[allow(dead_code)]
+    pub fn clear(&mut self) {
+        self.unique_entries.clear();
+        self.ordered_entries.clear();
+    }
+
+    /// Prune TXIDs that are older than `eviction_time` ago.
+    ///
+    // This method is public because ZIP-401 states about pruning:
+    // > This MAY be done periodically,
+    pub fn prune_old(&mut self) {
+        while let Some(txid) = self.front() {
+            let evicted_at = self
+                .unique_entries
+                .get(txid)
+                .expect("all entries should exist in both ordered_entries and unique_entries");
+            if self.has_expired(evicted_at) {
+                self.pop_front();
+            } else {
+                break;
+            }
+        }
+    }
+
+    /// Get the oldest TXID in the list.
+    fn front(&self) -> Option<&transaction::Hash> {
+        self.ordered_entries.front()
+    }
+
+    /// Removes the first element and returns it, or `None` if the `EvictionList`
+    /// is empty.
+    fn pop_front(&mut self) -> Option<transaction::Hash> {
+        if let Some(key) = self.ordered_entries.pop_front() {
+            let removed = self.unique_entries.remove(&key);
+            assert!(
+                removed.is_some(),
```

Comment:
Diagnostics Nitpick: Show the actual value.

```suggestion
            assert_eq!(
                removed,
                Some(key),
```